src/phoenix/db/insertion/dataset.py

```diff
@@ -0,0 +1,244 @@
+            span_rowid=span_rowid,
+            created_at=created_at,
+        )
+        .returning(models.DatasetExample.id)
+    )
+
+
+class RevisionKind(Enum):
+    CREATE = "CREATE"
+    PATCH = "PATCH"
+    DELETE = "DELETE"
+
+
+async def insert_dataset_example_revision(
+    session: AsyncSession,
+    dataset_version_id: DatasetVersionId,
+    dataset_example_id: DatasetExampleId,
+    input: Mapping[str, Any],
+    output: Mapping[str, Any],
+    metadata: Optional[Mapping[str, Any]] = None,
+    revision_kind: RevisionKind = RevisionKind.CREATE,
+    created_at: Optional[datetime] = None,
+) -> Optional[DatasetExampleRevisionId]:
+    return await session.scalar(
+        insert(models.DatasetExampleRevisions)
+        .values(
+            dataset_version_id=dataset_version_id,
+            dataset_example_id=dataset_example_id,
+            input=input,
+            output=output,
+            metadata_=metadata or {},
+            revision_kind=revision_kind.value,
+            created_at=created_at,
+        )
+        .returning(models.DatasetExampleRevisions.id)
+    )
+
+
+class DatasetTableAction(Enum):
+    CREATE = "create"
+    APPEND = "append"
+
+    @classmethod
+    def _missing_(cls, v: Any) -> Optional["DatasetTableAction"]:
+        if isinstance(v, str) and v and v.isascii() and not v.islower():
+            return cls(v.lower())
+        return None
+
+
+async def add_table(
+    session: AsyncSession,
+    table: Iterable[Mapping[str, Any]],
+    input_keys: Sequence[str],
+    output_keys: Sequence[str],
+    metadata_keys: Sequence[str] = (),
+    name: Optional[str] = None,
+    description: Optional[str] = None,
+    metadata: Optional[Mapping[str, Any]] = None,
+    action: DatasetTableAction = DatasetTableAction.CREATE,
+) -> Optional[DatasetCreationEvent]:
+    if not name:
+        if action is DatasetTableAction.CREATE:
+            name = await _get_unique_name(session)
+        else:
+            raise ValueError(f"Dataset name must not be empty for action={action.value}")
+    keys = DatasetKeys(frozenset(input_keys), frozenset(output_keys), frozenset(metadata_keys))
+    created_at = datetime.now(timezone.utc)
+    dataset_id: Optional[DatasetId] = None
+    if action is DatasetTableAction.APPEND and name:
+        dataset_id = await session.scalar(
+            select(models.Dataset.id).where(models.Dataset.name == name)
+        )
+    if action is DatasetTableAction.CREATE or dataset_id is None:
+        try:
+            dataset_id = await insert_dataset(
+                session=session,
+                name=name,
+                description=description,
+                metadata=metadata,
+                created_at=created_at,
+            )
+        except Exception:
+            logger.exception(
+                f"Fail to insert dataset: {input_keys=}, {output_keys=}, {metadata_keys=}"
+            )
+            raise
+    assert dataset_id is not None
```

Comment:
```suggestion
    assert dataset_id is not None, "Inserting dataset failed"
```

If we want to use asserts rather than raising errors, can we add a human-readable message in case the assert fails?

Reply:
these are really meant for type-checkers so they don't complain